# Online Retail Segmentation

In [2]:
# data analysis and manipulation tool
import pandas as pd

# perform numerical computation
import numpy as np

# data viz library
import matplotlib.pyplot as plt 
import seaborn as sns

In [3]:
df = pd.read_excel("Data/OnlineRetail.xlsx")

In [4]:
# display our data 
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

## Additional Variable Information

**InvoiceNo:** Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.

**StockCode:** Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.

**Description:** Product (item) name. Nominal.

**Quantity:** The quantities of each product (item) per transaction. Numeric.	

**InvoiceDate:** Invoice Date and time. Numeric, the day and time when each transaction was generated.

**UnitPrice:** Unit price. Numeric, Product price per unit in sterling.

**CustomerID:** Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.

**Country:** Country name. Nominal, the name of the country where each customer resides. 

In [5]:
# create a copy of our data that we will work with
retail_df = df.copy()

In [6]:
# display our copied data 
retail_df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [7]:
# Generate descriptive statistics.
retail_df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.081158                            NaN      96.759853   

          CustomerID  
count  406829.000000  
mean    15287.690570  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.600303

In [8]:
# Generate descriptive statistics for object type
retail_df.describe(include="O")

InvoiceNo StockCode                         Description  \
count      541909    541909                              540455   
unique      25900      4070                                4223   
top        573585    85123A  WHITE HANGING HEART T-LIGHT HOLDER   
freq         1114      2313                                2369   

               Country  
count           541909  
unique              38  
top     United Kingdom  
freq            495478

In [9]:
# Print a concise summary of a DataFrame.
retail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


**NOTE:** From the above,

the following columns contain data as numbers but they are of Dtype object: InvoiceNo, StockCode.

This shows there must be a reason for such Data types, i.e: they contain letters among numbers

# Exploratory Data Analysis

Analyze and investigate data sets and summarize their main characteristics, often employing data visualization methods.


In [10]:
retail_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

### Review each column data 

#### 1 -> InvoiceNo

In [11]:
# Check Its Dtype
retail_df["InvoiceNo"].dtype

dtype('O')

Above results

Question: why Dtype object yet the data seem to be of numbers?

Answer: Maybe it contains string data as well.

In [12]:
retail_df["InvoiceNo"] = retail_df["InvoiceNo"].astype("str") 

In [13]:
#Use regular expressions to identify string data

display(retail_df["InvoiceNo"].str.replace("[0-9]", "", regex=True).unique())
print("------------------------------------------")

print("show the value_count of each letter")
display(retail_df["InvoiceNo"].str.replace("[0-9]", "", regex=True).value_counts())

array(['', 'C', 'A'], dtype=object)

------------------------------------------
show the value_count of each letter


InvoiceNo
     532618
C      9288
A         3
Name: count, dtype: int64

Observation above:

Column InvoiceNo contains letters C and A as part of its data

#### 2 -> StockCode

In [14]:
# Column data type
retail_df["StockCode"].dtype

dtype('O')

In [15]:
retail_df["StockCode"] = retail_df["StockCode"].astype("str") 

In [16]:
# check if StockCode is a combination of numbers only or a mix of digits and letters
retail_df[(retail_df["StockCode"].str.match("^\\d{5}$") == False)]

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
45        536370      POST                              POSTAGE         3   
...          ...       ...                                  ...       ...   
541778    581578    84997C      CHILDRENS CUTLERY POLKADOT BLUE         8   
541809    581579    85099C       JUMBO  BAG BAROQUE BLACK WHITE        10   
541838    581580    84993A            75 GREEN PETIT FOUR CASES         2   
541844    581580    85049A        TRADITIONAL CHRISTMAS RIBBONS         1   
541853    581580    85049E            SCANDINAVIAN REDS RIBBONS         2   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
45     2010-12-01 08:45:00      18.00     12583.0          France  
...                    ...        ...         ...             ...  
541778 2011-12-09 12:16:00       4.15     12713.0         Germany  
541809 2011-12-09 12:19:00       1.79     17581.0  United Kingdom  
541838 2011-12-09 12:20:00       0.42     12748.0  United Kingdom  
541844 2011-12-09 12:20:00       1.25     12748.0  United Kingdom  
541853 2011-12-09 12:20:00       1.25     12748.0  United Kingdom  

[54873 rows x 8 columns]

In [17]:
#check if we have other patterns
retail_df[(retail_df["StockCode"].str.match("^\\d{5}$") == False) & (retail_df["StockCode"].str.match("^\\d{5}[a-zA-Z]+$") == False)]

InvoiceNo StockCode     Description  Quantity         InvoiceDate  \
45        536370      POST         POSTAGE         3 2010-12-01 08:45:00   
141      C536379         D        Discount        -1 2010-12-01 09:41:00   
386       536403      POST         POSTAGE         1 2010-12-01 11:27:00   
1123      536527      POST         POSTAGE         1 2010-12-01 13:04:00   
1423      536540        C2        CARRIAGE         1 2010-12-01 14:05:00   
...          ...       ...             ...       ...                 ...   
541540    581498       DOT  DOTCOM POSTAGE         1 2011-12-09 10:26:00   
541541   C581499         M          Manual        -1 2011-12-09 10:28:00   
541730    581570      POST         POSTAGE         1 2011-12-09 11:59:00   
541767    581574      POST         POSTAGE         2 2011-12-09 12:09:00   
541768    581578      POST         POSTAGE         3 2011-12-09 12:16:00   

        UnitPrice  CustomerID         Country  
45          18.00     12583.0          France  
141         27.50     14527.0  United Kingdom  
386         15.00     12791.0     Netherlands  
1123        18.00     12662.0         Germany  
1423        50.00     14911.0            EIRE  
...           ...         ...             ...  
541540    1714.17         NaN  United Kingdom  
541541     224.69     15498.0  United Kingdom  
541730      18.00     12662.0         Germany  
541767      18.00     12526.0         Germany  
541768      18.00     12713.0         Germany  

[2995 rows x 8 columns]

In [18]:
#check for unique patterns from above dataframe
retail_df[(retail_df["StockCode"].str.match("^\\d{5}$") == False) & (retail_df["StockCode"].str.match("^\\d{5}[a-zA-Z]+$") == False)]["StockCode"].unique()

array(['POST', 'D', 'C2', 'DOT', 'M', 'BANK CHARGES', 'S', 'AMAZONFEE',
       'DCGS0076', 'DCGS0003', 'gift_0001_40', 'DCGS0070', 'm',
       'gift_0001_50', 'gift_0001_30', 'gift_0001_20', 'DCGS0055',
       'DCGS0072', 'DCGS0074', 'DCGS0069', 'DCGS0057', 'DCGSSBOY',
       'DCGSSGIRL', 'gift_0001_10', 'PADS', 'DCGS0004', 'DCGS0073',
       'DCGS0071', 'DCGS0068', 'DCGS0067', 'DCGS0066P', 'B', 'CRUK'],
      dtype=object)

# Data Cleaning

In [37]:
# make a copy of data the EDA-data to be used in data cleaning 
cleaned_df = retail_df.copy()

#### 1 -> InvoiceNo

In [38]:
cleaned_df["InvoiceNo"].dtype

dtype('O')

In [39]:
cleaned_df["InvoiceNo"] = cleaned_df["InvoiceNo"].astype("str")

In [40]:
mask = (
    cleaned_df["InvoiceNo"].str.match("^\\d{6}$") == True
)

cleaned_df = cleaned_df[mask]

cleaned_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[532618 rows x 8 columns]

#### 2 -> StockCode

In [41]:
cleaned_df["StockCode"].dtype

dtype('O')

In [42]:
cleaned_df["StockCode"] = cleaned_df["StockCode"].astype("str")

In [43]:
# symbol | means a pipe that stands for or
mask = (
    (cleaned_df["StockCode"].str.match("^\\d{5}$") == True)
    | (cleaned_df["StockCode"].str.match("^\\d{5}[a-zA-Z]+$") == True)
)

cleaned_df = cleaned_df[mask]

cleaned_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[530210 rows x 8 columns]

**Remove null values**

In [44]:
cleaned_df.dropna(inplace=True)

C:\Users\user\AppData\Local\Temp\ipykernel_12996\3584568400.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_df.dropna(inplace=True)


**Reviewing our data**

In [45]:
cleaned_df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [46]:
cleaned_df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  396370.000000                         396370  396370.000000   
mean       13.046474  2011-07-11 00:04:30.985594624       2.867983   
min         1.000000            2010-12-01 08:26:00       0.000000   
25%         2.000000            2011-04-07 11:16:00       1.250000   
50%         6.000000            2011-07-31 14:39:00       1.950000   
75%        12.000000            2011-10-20 14:41:00       3.750000   
max     80995.000000            2011-12-09 12:50:00     649.500000   
std       180.733597                            NaN       4.264566   

          CustomerID  
count  396370.000000  
mean    15301.602071  
min     12346.000000  
25%     13975.000000  
50%     15159.000000  
75%     16803.000000  
max     18287.000000  
std      1709.982593

Observations above

The minimum price appears to be 0, which will not be used in our analysis

Check the length of unit price at 0 and then drop the corresponding row

In [47]:
len(cleaned_df[cleaned_df["UnitPrice"] == 0])

33

In [48]:
cleaned_df = cleaned_df[cleaned_df["UnitPrice"] > 0]

In [49]:
cleaned_df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  396337.000000                         396337  396337.000000   
mean       13.012547  2011-07-11 00:04:38.980463616       2.868222   
min         1.000000            2010-12-01 08:26:00       0.040000   
25%         2.000000            2011-04-07 11:16:00       1.250000   
50%         6.000000            2011-07-31 14:39:00       1.950000   
75%        12.000000            2011-10-20 14:41:00       3.750000   
max     80995.000000            2011-12-09 12:50:00     649.500000   
std       179.641245                            NaN       4.264663   

          CustomerID  
count  396337.000000  
mean    15301.689416  
min     12346.000000  
25%     13975.000000  
50%     15159.000000  
75%     16803.000000  
max     18287.000000  
std      1709.961948

**Get the percentage of data we are left with after cleaning**

In [60]:
print( f"{round( (len(cleaned_df)/len(df))*100, 2)}%" )

73.14%


We are remaining with 73.14% of data to perform our analysis

# Feature Engineering